### PR TITLE
C# Exclude dynamic casts from useless casts check

### PR DESCRIPTION
**Pull request by tamasvajk**
This PR removes all casts to `dynamic` from the useless upcast check. The issue was reported in https://github.com/github/codeql/issues/8069.

### CODE REVIEW

File: csharp/ql/src/Language Abuse/UselessUpcast.ql [hunk 1]
```diff
@@ -204,6 +204,19 @@ class ExplicitUpcast extends ExplicitCast {
     this = any(LocalVariableDeclAndInitExpr decl | decl.isImplicitlyTyped()).getInitializer()
     or
     exists(LambdaExpr c | c.canReturn(this))
+    or
+    dest instanceof DynamicType
+  }
+}
+
+// Stop external filepaths from appearing in the results
+class ValueOrRefTypeLocation extends ValueOrRefType {
```
Comment (michaelnebel) [inline csharp/ql/src/Language Abuse/UselessUpcast.ql:213]: Why is this needed?

**Review by michaelnebel — APPROVED** [review]
Looks plausible to me!

File: csharp/ql/src/Language Abuse/UselessUpcast.ql (hunk 1 above)
Comment (tamasvajk) [reply csharp/ql/src/Language Abuse/UselessUpcast.ql:213]: Ahh, thanks. This should not be added, especially not here (it would only have a place in tests). In the test we're listing types, which are not necessarily `fromSource`. This means that their path would point to framework locations, which is platform and installation dependent. 

This change is not needed, because we exclude the only framework type that would show up in the tests. See [here](https://github.com/github/codeql/pull/8152/commits/7cba3c59515017d322e2256c0ce8a0d90338deae#diff-afbeb8fdd27e6e411446f8b13e44a33c1f08a527498d1cf1aa96ad84bc502cebL9).

**Review by hvitved — APPROVED** [review]

**Review by michaelnebel — APPROVED** [review]
Looks good to me!
